db/db_options_test.cc

```diff
@@ -530,17 +534,218 @@ TEST_F(DBOptionsTest, RunStatsDumpPeriodSec) {
   rocksdb::SyncPoint::GetInstance()->EnableProcessing();
   Reopen(options);
   ASSERT_EQ(5, dbfull()->GetDBOptions().stats_dump_period_sec);
-  dbfull()->TEST_WaitForTimedTaskRun([&] { mock_env->set_current_time(5); });
+  dbfull()->TEST_WaitForDumpStatsRun([&] { mock_env->set_current_time(5); });
   ASSERT_GE(counter, 1);
 
   // Test cacel job through SetOptions
   ASSERT_OK(dbfull()->SetDBOptions({{"stats_dump_period_sec", "0"}}));
   int old_val = counter;
-  env_->SleepForMicroseconds(10000000);
+  for (int i = 6; i < 20; ++i) {
+    mock_env->set_current_time(i);
+  }
   ASSERT_EQ(counter, old_val);
   Close();
 }
 
+// Test persistent stats background thread scheduling and cancelling
+TEST_F(DBOptionsTest, StatsPersistScheduling) {
+  Options options;
+  options.create_if_missing = true;
+  options.stats_persist_period_sec = 5;
+  std::unique_ptr<rocksdb::MockTimeEnv> mock_env;
+  mock_env.reset(new rocksdb::MockTimeEnv(env_));
+  mock_env->set_current_time(0);  // in seconds
+  options.env = mock_env.get();
+  int counter = 0;
+  rocksdb::SyncPoint::GetInstance()->SetCallBack(
+      "DBImpl::PersistStats:Entry", [&](void* /*arg*/) { counter++; });
+  rocksdb::SyncPoint::GetInstance()->EnableProcessing();
+  Reopen(options);
+  ASSERT_EQ(5, dbfull()->GetDBOptions().stats_persist_period_sec);
+  dbfull()->TEST_WaitForPersistStatsRun([&] { mock_env->set_current_time(5); });
+  ASSERT_GE(counter, 1);
+
+  // Test cacel job through SetOptions
+  ASSERT_OK(dbfull()->SetDBOptions({{"stats_persist_period_sec", "0"}}));
+  int old_counter = counter;
+  for (int i = 6; i < 20; ++i) {
+    mock_env->set_current_time(i);
+  }
+  ASSERT_EQ(counter, old_counter);
+
+  // Test resume job through SetOptions
+  ASSERT_OK(dbfull()->SetDBOptions({{"stats_persist_period_sec", "5"}}));
+  ASSERT_EQ(5, dbfull()->GetDBOptions().stats_persist_period_sec);
+  for (int i = 20; i < 40; ++i) {
+    mock_env->set_current_time(i);
+  }
+  ASSERT_GE(counter, old_counter);
```

Comment:
Why is it ASSERT_GE(), not ASSERT_GT()?

I also don't understand why the test can pass. mock_env->set_current_time() doesn't wake up anything, and we don't wait for anything before the checking. What make sure the counter is advanced?

Reply:
will update with ASSERT_GT in the next upload. 
In this test we didn't call `TEST_WaitForPersistStatsRun` so `RepeatableThread::wait` simply checks if the current time has passed the `wait_until`, see
https://github.com/facebook/rocksdb/blob/master/util/repeatable_thread.h#L95

Reply:
This seems like a flaky test and since the part where we set stats_persist_period_sec from 0->5, hoping that the persist stats will run in the background is already covered in the next test `DBOptionsTest.PersistentStatsFreshInstall`, it will be removed in the next upload